reward/supply_manager.go

```diff
@@ -0,0 +1,377 @@
+	// Externally injected dependencies
+	chain              blockChain
+	chainHeadChan      chan blockchain.ChainHeadEvent
+	chainHeadSub       event.Subscription
+	gov                governanceHelper
+	db                 database.DBManager
+	checkpointInterval uint64
+
+	// Internal data structures
+	accRewardCache *lru.ARCCache  // Cache (number uint64) -> (accReward *database.AccReward)
+	memoCache      *lru.ARCCache  // Cache (address Address) -> (memo.Burnt *big.Int)
+	quit           uint32         // Stop the goroutine in initial catchup stage
+	quitCh         chan struct{}  // Stop the goroutine in event subscription state
+	wg             sync.WaitGroup // background goroutine wait group for shutting down
+}
+
+// NewSupplyManager creates a new supply manager.
+// The TotalSupply data is stored every checkpointInterval blocks.
+func NewSupplyManager(chain blockChain, gov governanceHelper, db database.DBManager, checkpointInterval uint) *supplyManager {
+	accRewardCache, _ := lru.NewARC(supplyCacheSize)
+	memoCache, _ := lru.NewARC(10)
+
+	return &supplyManager{
+		chain:              chain,
+		chainHeadChan:      make(chan blockchain.ChainHeadEvent, chainHeadChanSize),
+		gov:                gov,
+		db:                 db,
+		checkpointInterval: uint64(checkpointInterval),
+		accRewardCache:     accRewardCache,
+		memoCache:          memoCache,
+		quitCh:             make(chan struct{}, 1), // make sure Stop() doesn't block if catchup() has exited before Stop()
+	}
+}
+
+func (sm *supplyManager) Start() {
+	sm.wg.Add(1)
+	go sm.catchup()
+}
+
+func (sm *supplyManager) Stop() {
+	atomic.StoreUint32(&sm.quit, 1)
+	sm.quitCh <- struct{}{}
+	sm.wg.Wait()
+	if sm.chainHeadSub != nil {
+		sm.chainHeadSub.Unsubscribe()
+	}
+}
+
+func (sm *supplyManager) GetAccReward(num uint64) (*database.AccReward, error) {
+	if accReward, ok := sm.accRewardCache.Get(num); ok {
+		return accReward.(*database.AccReward), nil
+	}
+
+	if accReward := sm.db.ReadAccReward(num); accReward != nil {
+		sm.accRewardCache.Add(num, accReward.Copy())
+		return accReward, nil
+	}
+
+	return nil, errNoAccReward
+}
+
+func (sm *supplyManager) GetTotalSupply(num uint64) (*TotalSupply, error) {
+	ts := &TotalSupply{}
+
+	// Read accumulated rewards (minted, burntFee)
+	accReward, err := sm.GetAccReward(num)
+	if err != nil {
+		return nil, err
+	}
+	ts.TotalMinted = accReward.Minted
+	ts.BurntFee = accReward.BurntFee
+
+	// Read canonical burn address balances
+	header := sm.chain.GetHeaderByNumber(num)
+	if header == nil {
+		return nil, errNoBlock
+	}
+	state, err := sm.chain.StateAt(header.Root)
+	if err != nil {
+		return nil, err
+	}
+	ts.ZeroBurn = state.GetBalance(zeroBurnAddress)
+	ts.DeadBurn = state.GetBalance(deadBurnAddress)
+
+	// Read KIP103 and KIP160 burns
+	bigNum := new(big.Int).SetUint64(num)
+	config := sm.chain.Config()
+
+	ts.Kip103Burn = big.NewInt(0)
+	if config.Kip103CompatibleBlock != nil && config.Kip103CompatibleBlock.Cmp(bigNum) <= 0 {
```

Comment:
```suggestion
	if config.IsKIP103ForkBlock(bigNum) {
```

Reply:
No, the statement checks if `kip103block <= num`. The `IsKIP103ForkBlock` function checks if `kip103block == num`. Oh the other hand, it's inappropriate to add `IsKIP103ForkEnabled()` because KIP103 is an one-time action.

Reply:
Oops, my bad. Sorry for buzzing.